module/VuFind/src/VuFind/Auth/ChoiceAuth.php

```diff
@@ -340,7 +340,9 @@ protected function proxyAuthMethod($method, $params)
             return false;
         }
 
-        if (!in_array($this->strategy, $this->strategies)) {
+        if ('Email' !== $this->strategy
```

Comment:
Is there a reason why we need a special case for Email at this point? Is the idea that other methods can turn into Email even if it's not configured as a top-level option?

Reply:
Yes, that's the idea. I'll at least clarify it with a comment. I'll also add a hash check to prevent abuse, but that's a bit hacky since Manager needs to know about how Email behaves, so I'm open to better suggestions.

Reply:
If we can figure out a new method to add to the authentication methods to better abstract this special case, we might be able to refactor to something like this:

<pre>
        if (!$this->getPluginManager()->has($this->strategy)) {
            throw new InvalidArgumentException("Unknown setting: {$this->strategy}");
        }
        $authenticator = $this->getPluginManager()->get($this->strategy);
        $authenticator->setConfig($this->getConfig());
        if (!$authenticator->spawnedBy($this->strategies)
            && !in_array($this->strategy, $this->strategies)
        ) {
            throw new InvalidArgumentException("Illegal setting: {$this->strategy}");
        }
</pre>

I don't claim that `spawnedBy` is a good method name, or that this is exactly the best approach... but my thought is that if an authentication method can be triggered by a different method, it could have a method that looks at the array of strategies and decides whether or not it's valid to use that method. In the case of email, perhaps this could also check a flag within the session, so it only returns true when it knows it is valid, not just because it might theoretically be valid.

Does that make any sense?

Reply:
I came up with similar but slightly different way of asking the auth method if it wants to delegate control. What do you think? I also added email auth support to Demo driver to make it easier to test.